libnetdata/dictionary/dictionary.c

```diff
@@ -399,105 +724,229 @@ static inline size_t reference_counter_free(DICTIONARY *dict) {
+        refcount = --item->refcount;
     }
+    else {
+        // get the flags before decrementing any reference counters
+        // (the other way around may lead to use-after-free)
+        is_deleted = item_flag_check(item, ITEM_FLAG_DELETED);
 
-    if(refcount == 1) {
-        if((nv->flags & NAME_VALUE_FLAG_DELETED))
-            DICTIONARY_STATS_PENDING_DELETES_PLUS1(dict);
+        // decrement the refcount
+        refcount = __atomic_sub_fetch(&item->refcount, 1, __ATOMIC_SEQ_CST);
+    }
+
+    if(refcount < 0) {
+        internal_error(
+            true,
+            "DICTIONARY: attempted to release item without references (refcount = %d): "
+            "'%s' on dictionary created by %s() (%zu@%s)",
+            refcount + 1,
+            item_get_name(item),
+            dict->creation_function,
+            dict->creation_line,
+            dict->creation_file);
+
+        fatal(
+            "DICTIONARY: attempted to release item '%s' without references (refcount = %d)",
+            item_get_name(item),
+            refcount + 1);
+    }
+
+    if(refcount == 0) {
+
+        if(is_deleted)
+            DICTIONARY_PENDING_DELETES_PLUS1(dict);
 
         // referenced items counts number of unique items referenced
         // so, we decrease it only when refcount == 0
-        DICTIONARY_STATS_REFERENCED_ITEMS_MINUS1(dict);
+        DICTIONARY_REFERENCED_ITEMS_MINUS1(dict);
     }
+}
 
-    if(can_get_write_lock && DICTIONARY_STATS_PENDING_DELETES_GET(dict)) {
-        // we can garbage collect now
+static int item_check_and_acquire_advanced(DICTIONARY *dict, DICTIONARY_ITEM *item, bool having_index_lock) {
+    size_t spins = 0;
+    REFCOUNT refcount, desired;
 
-        dictionary_lock(dict, DICTIONARY_LOCK_WRITE);
-        garbage_collect_pending_deletes_unsafe(dict);
-        dictionary_unlock(dict, DICTIONARY_LOCK_WRITE);
-    }
+    do {
+        spins++;
 
-    return refcount;
-}
+        refcount = DICTIONARY_ITEM_REFCOUNT_GET(dict, item);
 
-// ----------------------------------------------------------------------------
-// hash table
+        if(refcount < 0) {
+            // we can't use this item
+            return ITEM_IS_CURRENTLY_BEING_DELETED;
+        }
 
-static void hashtable_init_unsafe(DICTIONARY *dict) {
-    dict->JudyHSArray = NULL;
-}
+        if(item_flag_check(item, ITEM_FLAG_DELETED)) {
+            // we can't use this item
+            return ITEM_MARKED_FOR_DELETION;
+        }
 
-static size_t hashtable_destroy_unsafe(DICTIONARY *dict) {
-    if(unlikely(!dict->JudyHSArray)) return 0;
+        desired = refcount + 1;
 
-    JError_t J_Error;
-    Word_t ret = JudyHSFreeArray(&dict->JudyHSArray, &J_Error);
-    if(unlikely(ret == (Word_t) JERR)) {
-        error("DICTIONARY: Cannot destroy JudyHS, JU_ERRNO_* == %u, ID == %d",
-              JU_ERRNO(&J_Error), JU_ERRID(&J_Error));
-    }
+    } while(!__atomic_compare_exchange_n(&item->refcount, &refcount, desired,
```

Comment:
@ktsaou @stelfrag  I recall us discussing the performance of this piece of code and I found this interesting https://youtu.be/rmGJc9PXpuE?t=16